src/auto_archiver/modules/opentimestamps_enricher/opentimestamps_enricher.py

```
@@ -0,0 +1,174 @@
+                        logger.debug(f"Successfully submitted to calendar: {calendar.url}")
+                        submitted_to_calendar = True
+                    except Exception as e:
+                        logger.warning(f"Failed to submit to calendar {calendar.url}: {e}")
+                
+                # If all calendar submissions failed, add pending attestations
+                if not submitted_to_calendar and not timestamp.attestations:
+                    logger.error(f"Failed to submit to any calendar for {file_path}. **This file will not be timestamped.**")
+                    media.set("opentimestamps", False)
+                    continue
+                
+                # Save the timestamp proof to a file
+                timestamp_path = os.path.join(self.tmp_dir, f"{os.path.basename(file_path)}.ots")
+                try:
+                    with open(timestamp_path, 'wb') as f:
+                        # Create a serialization context and write to the file
+                        ctx = serialize.BytesSerializationContext()
+                        detached_timestamp.serialize(ctx)
+                        f.write(ctx.getbytes())
+                except Exception as e:
+                    logger.warning(f"Failed to serialize timestamp file: {e}")
+                    continue
+                
+                # Create media for the timestamp file
+                timestamp_media = Media(filename=timestamp_path)
+                # explicitly set the mimetype, normally .ots files are 'application/vnd.oasis.opendocument.spreadsheet-template'
+                timestamp_media.mimetype = "application/vnd.opentimestamps"
+                timestamp_media.set("opentimestamps_version", opentimestamps.__version__)
+                
+                verification_info = self.verify_timestamp(detached_timestamp)
+                for key, value in verification_info.items():
+                    timestamp_media.set(key, value)
```

Would this store failures from the verification too?

---

I'm not following, this will store either 'status: pending' or 'status: confirmed' if the timestamp was submitted. If no timestamp was submitted successfully, then this won't be set.

---

Ah okay, is it not possible for there to be no attestations at that point? aka if it's submitted there's always attestations, then these would be caught by `if not submitted_to_calendar and not timestamp.attestations:`

Does it need the conditional in that case?
```   
        if attestations:
            ...
        else:
            result["verified"] = False
```

(maybe I am misunderstanding the process, please ignore me if so!)

---

> Ah okay, is it not possible for there to be no attestations at that point?

FWIW the `.ots` file format requires timestamp proofs to have at least one attestation; it's syntactically invalid if they don't, and the python-opentimestamps libraries will give you an exception. Specifically, it'll be an IOError as from the parser's point of view, the data is truncated.

---

Yes, correct. attestations means at least 1 successful submit was made.

If there are no attestations, then there's nothing to verify, and 'opentimestamped' on the parent Meta would explicitly be set as 'False'.(see line 83)

---

Would this be redundant code, maybe we should remove it as it would hide unexpected errors?

        else:
            result["verified"] = False